stl/inc/atomic

```diff
@@ -118,6 +117,9 @@ _NODISCARD extern "C" bool __cdecl __std_atomic_has_cmpxchg16b() noexcept;
 #define ATOMIC_LLONG_LOCK_FREE    2
 #define ATOMIC_POINTER_LOCK_FREE  2
 
+_EXTERN_C
+_Smtx_t* __stdcall __std_atomic_get_mutex(const void* _Key) noexcept;

```

Comment:
Do you consider being able to change the lock that this uses in the future a useful ability? If these were lock and unlock functions we could do that in a future DLL update.

No change requested.

Reply:
We can:
1. Stop storing pointer to the lock, always locate it. This will lead to unused lock member in non-lock-free `atomic_ref` that can be avoided under new ABI (`[[no_unique_address]]`). Note that it can't be avoided by template specialization.
2. Store mutex, but still call DLL lock function. Though we'll lose potential aliasobj for `<shared_mutex>`. We can use forwarders instead though. See #763

Reply:
#763 doesn't seem super relevant because if we used lock and unlock functions here we would be calling the Windows APIs from our separate compiled world directly, no need for forwarders then.

Reply:
There's an option to use lock and unlock, and still keep get_mutex returning an opaque pointer.

Reply:
I gave this one more thought. All options make sense, including:
1. Having `__std_atomic_get_mutex` only and calling SRWLOCK functions
2. Having `__std_atomic_get_mutex` returning an opaque pointer and calling implementation functions that can be changed later
3. Having only implementation functions, make `__std_atomic_get_mutex` internal.

The later option will cause dummy member to exist in current `atomic_ref` due to ABI limitation. First two utilize it as dummy pointer.

The first one is eligible for `aliasobj` optimization. The second is not eligible for aliasobj optimization, but for `/MD` can use forwarders optimization.

Right now first option is used, but I can change it. @BillyONeal , what do you think?

Reply:
If I could do this:
```C++
#include <type_traits>

template<bool IsRef>
struct A
{
    void* p;
    std::conditional_t<IsRef, void(void), void*> l;
};

int main()
{
    static_assert(sizeof(A<false>) == sizeof(void*) * 2, "ptr sizemismatch");
    static_assert(sizeof(A<true>) == sizeof(void*), "ptr sizemismatch");
}
```
The third option would have been a bit more appealing

Reply:
I just wanted to make sure the separate lock/unlock function design was considered and that you are OK with baking the SRWLOCK assumption into ABI. I'm happy with your P/R as written.